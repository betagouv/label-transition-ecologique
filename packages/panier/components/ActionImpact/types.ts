import {
  ActionImpactFourchetteBudgetaire,
  ActionImpactTempsMiseEnOeuvre,
  ActionImpactThematique,
} from '@tet/api';

export type ActionImpactProps = {
  /** Titre de l'action à impact */
  titre: string;
  /** Thématiques de l'action à impact */
  thematiques: ActionImpactThematique[];
  /** Budget de la mise en place de l'action : petit, moyen ou élevé */
  budget?: ActionImpactFourchetteBudgetaire;
  /** Description de l'action à impact */
  description: string;
  /** Temps de mise en oeuvre */
  miseEnOeuvre?: ActionImpactTempsMiseEnOeuvre;
  /** Lien vers les ressources externes */
  ressources?: string | null;
  /** Lien vers les retours d'expérience */
  rex?: string | null;
  /** Lien vers les subventions mobilisables */
  subventions?: string | null;
  /** Statut de l'action */
  statut?: 'non_pertinent' | 'en_cours' | 'realise' | null;
  /** Indique si l'action est mise dans le panier */
  panier: boolean;
  /** Initialisation de l'état sélectionné de la carte */
  isSelected?: boolean;
  /** Détecte le changement de statut sélectionné ou non */
  onToggleSelected: (value: boolean) => void;
  /** Détecte le changement de statut de l'action : en cours, réalisé */
  onUpdateStatus?: (status: string) => void;
};

export type ModaleActionImpactProps = {
  /** Composant enfant (carte action) */
  children: JSX.Element;
} & ActionImpactProps;

export type CarteActionImpactProps = Pick<
  ActionImpactProps,
  | 'titre'
  | 'thematiques'
  | 'budget'
  | 'statut'
  | 'panier'
  | 'isSelected'
  | 'onToggleSelected'
  | 'onUpdateStatus'
>;
