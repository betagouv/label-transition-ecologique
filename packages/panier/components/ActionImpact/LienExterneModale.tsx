import {Icon} from '@tet/ui';

type LienExterneModaleProps = {
  url: string;
  label: string;
};

const LienExterneModale = ({url, label}: LienExterneModaleProps) => {
  return (
    <div className="flex items-start gap-2">
      <Icon icon="external-link-line" size="lg" className="text-primary-4" />
      <a
        href={url}
        className="text-primary-10 text-sm leading-6 font-bold hover:text-primary-8 bg-none active:!bg-transparent after:hidden underline underline-offset-4"
        target="_blank"
        rel="noreferrer noopener"
      >
        {label}
      </a>
    </div>
  );
};

export default LienExterneModale;
