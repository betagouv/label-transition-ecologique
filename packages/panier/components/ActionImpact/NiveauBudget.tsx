import {ActionImpactFourchetteBudgetaire} from '@tet/api';
import {Icon, Tooltip} from '@tet/ui';
import classNames from 'classnames';

type NiveauBudgetProps = {
  budget: ActionImpactFourchetteBudgetaire;
};

const NiveauBudget = ({budget}: NiveauBudgetProps) => {
  const tooltipText = (
    <div className="flex flex-col gap-2 text-primary-10">
      <div className="font-bold">{budget.nom}</div>
      <div>
        Estimation budgétaire HT (investissement et fonctionnement, hors
        subvention).
      </div>
      <div>
        Une évaluation précise du budget sera à réaliser lors du dimensionnement
        exacte de l’action.
      </div>
    </div>
  );

  return (
    <Tooltip
      label={
        <div className="w-52 !font-normal">
          {budget.niveau === 1
            ? 'Ordre de grandeur budgétaire non estimé.'
            : tooltipText}
        </div>
      }
    >
      <div className="flex items-center justify-center px-1.5 py-1 border rounded border-grey-3 bg-grey-1">
        <Icon
          icon="money-euro-circle-fill"
          className={classNames({
            'text-secondary-1': budget.niveau > 1,
            'text-grey-4': budget.niveau === 1,
          })}
        />
        <Icon
          icon="money-euro-circle-fill"
          className={classNames({
            'text-secondary-1': budget.niveau > 2,
            'text-grey-4': budget.niveau <= 2,
          })}
        />
        <Icon
          icon="money-euro-circle-fill"
          className={classNames({
            'text-secondary-1': budget.niveau > 3,
            'text-grey-4': budget.niveau <= 3,
          })}
        />
      </div>
    </Tooltip>
  );
};

export default NiveauBudget;
