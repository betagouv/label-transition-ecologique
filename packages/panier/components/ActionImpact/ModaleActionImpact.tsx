/* eslint-disable react/no-unescaped-entities */
import classNames from 'classnames';
import {ModaleActionImpactProps} from './types';
import {
  Badge,
  Button,
  Divider,
  InfoTooltip,
  Modal,
  ModalFooter,
  ModalFooterSection,
} from '@tet/ui';
import LienExterneModale from './LienExterneModale';
import Markdown from '@tet/panier/components/Markdown';

const URL_AIDES_TERRITOIRES = 'https://aides-territoires.beta.gouv.fr';

/**
 * Modale action à impact du panier d'actions
 */

export const ModaleActionImpact = ({
  children,
  titre,
  thematiques,
  budget,
  description,
  miseEnOeuvre,
  ressources,
  rex,
  subventions,
  statut,
  panier,
  onToggleSelected,
  onUpdateStatus,
}: ModaleActionImpactProps) => {
  return (
    <Modal
      size="lg"
      title={titre}
      textAlign="left"
      render={() => {
        return (
          <div>
            {/* Badges thématiques */}
            {!!thematiques.length && (
              <div className="flex gap-4 flex-wrap mb-6">
                {thematiques.map(theme => (
                  <Badge
                    key={theme.id}
                    title={theme.nom}
                    size="sm"
                    state="standard"
                  />
                ))}
              </div>
            )}

            {/* Description */}
            <Markdown
              content={description}
              className="paragraphe-18 mb-8 [&_ul]:list-disc [&_ul]:pl-8"
            />

            {/* Temps de mise en oeuvre */}
            <p className="text-base text-primary-10 font-bold mb-8">
              Temps de mise en oeuvre :{' '}
              <span className="text-primary-8">
                {miseEnOeuvre?.nom.toLowerCase() ?? 'non estimé'}
              </span>
              <InfoTooltip
                activatedBy="click"
                iconClassName="ml-2"
                label={
                  <div className="w-52 !font-normal flex flex-col gap-3 p-1">
                    <div>
                      Temps estimatif correspondant au déploiement de l’action
                      une fois celle-ci validée, de son démarrage à ses
                      premières réalisations.
                    </div>
                    <div>
                      Cette temporalité peut varier en fonction des
                      priorisations, des moyens et des ressources disponibles.
                    </div>
                    <Button
                      variant="underlined"
                      size="xs"
                      className="!text-info-1 !border-info-1"
                      href="https://aide.territoiresentransitions.fr/fr/article/en-savoir-plus-sur-les-criteres-du-panier-dactions-a-impact-b37tky/#1-temps-de-mise-en-oeuvre"
                      external
                    >
                      En savoir plus
                    </Button>
                  </div>
                }
              />
            </p>

            {/* Estimation budgétaire */}
            <p className="text-base text-primary-10 font-bold mb-8">
              Estimation budgétaire :{' '}
              <span className="text-primary-8">
                {budget?.nom.toLowerCase() ?? 'non estimé'}
              </span>
              <InfoTooltip
                activatedBy="click"
                iconClassName="ml-2"
                label={
                  <div className="w-52 !font-normal flex flex-col gap-3 p-1">
                    <div>
                      Estimation budgétaire HT (investissement et
                      fonctionnement, hors subvention).
                    </div>
                    <div>
                      Une évaluation précise du budget sera à réaliser lors du
                      dimensionnement exacte de l’action.
                    </div>
                    <Button
                      variant="underlined"
                      size="xs"
                      className="!text-info-1 !border-info-1"
                      href="https://aide.territoiresentransitions.fr/fr/article/en-savoir-plus-sur-les-criteres-du-panier-dactions-a-impact-b37tky/#1-estimation-des-ordres-de-grandeur-budgetaires"
                      external
                    >
                      En savoir plus
                    </Button>
                  </div>
                }
              />
            </p>

            {/* Ressources externes */}
            {!!ressources && ressources.length > 0 && (
              <div className="flex gap-x-8 gap-y-3 flex-wrap pb-8">
                {ressources.map(r => (
                  <LienExterneModale key={r.label} {...r} />
                ))}
              </div>
            )}

            <Divider className="mt-4" />

            {/* REX */}
            <div className="mt-2">
              <h6 className="text-primary-10 text-base font-bold">
                D’autres collectivités l’ont fait :
              </h6>
              <div className="flex gap-x-8 gap-y-3 flex-wrap pb-8">
                {!rex || rex.length === 0 ? (
                  <span className="text-sm text-grey-7">
                    Exemples d'autres collectivités à venir
                  </span>
                ) : (
                  rex.map(r => <LienExterneModale key={r.label} {...r} />)
                )}
              </div>
            </div>

            {/* Subventions mobilisables */}
            <div className="mt-2">
              <h6 className="text-primary-10 text-base font-bold">
                Subventions mobilisables :{' '}
                <InfoTooltip
                  activatedBy="click"
                  iconClassName="ml-1"
                  label={
                    <div className="w-52 !font-normal flex flex-col gap-3 p-1">
                      <div>
                        De nombreux programmes d’aides nationaux et locaux
                        peuvent vous permettre de financer ou d’accompagner vos
                        projets.
                      </div>
                      <div>
                        Les liens directs vers les programmes concernés seront
                        prochainement disponibles sur la plateforme.
                      </div>
                    </div>
                  }
                />
              </h6>
              <div className="flex gap-x-8 gap-y-3 flex-wrap">
                <LienExterneModale
                  label="Aides territoires"
                  url={URL_AIDES_TERRITOIRES}
                />
                {!!subventions &&
                  subventions.length > 0 &&
                  subventions.map(
                    s =>
                      s.url &&
                      s.url !== URL_AIDES_TERRITOIRES &&
                      s.url !== `${URL_AIDES_TERRITOIRES}/` && (
                        <LienExterneModale key={s.label} {...s} />
                      ),
                  )}
              </div>
            </div>
          </div>
        );
      }}
      renderFooter={({close}) => (
        <ModalFooter variant="space">
          <ModalFooterSection>
            <Button
              variant="outlined"
              className={classNames({
                'bg-primary-2': statut === 'en_cours',
              })}
              onClick={() => {
                onUpdateStatus?.('en_cours');
                close();
              }}
            >
              En cours
            </Button>
            <Button
              variant="outlined"
              className={classNames({
                'bg-primary-2': statut === 'realise',
              })}
              onClick={() => {
                onUpdateStatus?.('realise');
                close();
              }}
            >
              Réalisée
            </Button>
          </ModalFooterSection>
          <Button
            icon={panier ? 'file-reduce-fill' : 'file-add-fill'}
            onClick={() => {
              onToggleSelected(!panier);
              close();
            }}
          >
            {panier ? 'Retirer du panier' : 'Ajouter au panier'}
          </Button>
        </ModalFooter>
      )}
    >
      {children}
    </Modal>
  );
};
