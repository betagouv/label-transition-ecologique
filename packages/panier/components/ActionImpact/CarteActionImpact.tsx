import classNames from 'classnames';
import {CarteActionImpactProps} from './types';
import {Badge, Button, Card} from '@tet/ui';
import NiveauBudget from './NiveauBudget';

/**
 * Carte action à impact du panier d'actions
 */

export const CarteActionImpact = ({
  titre,
  thematiques,
  budget,
  panier,
  isSelected,
  statut,
  onToggleSelected,
  onUpdateStatus,
}: CarteActionImpactProps) => {
  const handleToggleSelect = (value: boolean) => onToggleSelected(value);

  return (
    <Card
      className={classNames('box-content px-4 py-4 h-28', {
        '!cursor-default': panier && !isSelected,
      })}
      isSelected={isSelected}
      footer={
        <div className="relative z-0">
          <div className="flex justify-end items-center gap-2 opacity-100 group-hover:opacity-0 transition-opacity duration-500">
            {/* Badge thématique */}
            {!!thematiques.length && (
              <Badge title={thematiques[0].nom} size="sm" state="standard" />
            )}
            {/* Budget */}
            <NiveauBudget budget={budget ?? {niveau: 4, nom: 'Non estimé'}} />
          </div>

          {/* Boutons d'action, visibles au hover de la carte */}
          <div className="opacity-0 group-hover:opacity-100 transition-opacity duration-500 w-full absolute z-10 bottom-0 right-0 bg-white flex justify-end gap-2">
            {panier ? (
              <Button size="xs" onClick={() => handleToggleSelect(false)}>
                Retirer du panier
              </Button>
            ) : (
              <>
                <Button
                  variant="outlined"
                  size="xs"
                  className={classNames({
                    '!bg-primary-2': statut === 'en_cours',
                  })}
                  onClick={() => onUpdateStatus?.('en_cours')}
                >
                  En cours
                </Button>
                <Button
                  variant="outlined"
                  size="xs"
                  className={classNames({
                    '!bg-primary-2': statut === 'realise',
                  })}
                  onClick={() => onUpdateStatus?.('realise')}
                >
                  Réalisée
                </Button>

                <Button size="xs" onClick={() => handleToggleSelect(true)}>
                  Ajouter au panier
                </Button>
              </>
            )}
          </div>
        </div>
      }
    >
      <span className="line-clamp-3">{titre}</span>
    </Card>
  );
};
