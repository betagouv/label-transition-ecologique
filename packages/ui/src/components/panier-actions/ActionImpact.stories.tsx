import {Meta, StoryObj} from '@storybook/react';
import {ActionImpact} from './ActionImpact';

const meta: Meta<typeof ActionImpact> = {
  title: "Composants métier/Panier d'actions/Action à impact",
  component: ActionImpact,
  argTypes: {
    complexite: {
      control: {type: 'select'},
    },
    budget: {
      control: {type: 'select'},
    },
  },
  args: {
    titre:
      'Désigner une (ou plusieurs personnes) chef de projet transition écologique',
    categorie: 'Mieux consommer',
    complexite: 1,
    budget: 2,
    description:
      "Un comité de pilotage se réunit régulièrement (au moins 2 fois par an) pour s'assurer du bon avancement et de la priorisation de la programmation des actions de transition écologique. Ce comité de pilotage s'assure que les actions de transition écologique sont identifiées et priorisées dans l'exercice de programmation budgétaire et la programmation pluriannuelle d'investissement",
    ressources: 'https://www.territoiresentransitions.fr/',
    nbCollectivitesEnCours: 12,
    nbCollectivitesRealise: 8,
    onToggleSelected: () => {},
    onUpdateStatus: () => {},
  },
};

export default meta;

type Story = StoryObj<typeof ActionImpact>;

export const Default: Story = {};
