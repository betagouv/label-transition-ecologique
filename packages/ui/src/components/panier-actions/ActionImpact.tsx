import {useEffect, useState} from 'react';
import {CarteActionImpact} from './CarteActionImpact';
import {ModaleActionImpact} from './ModaleActionImpact';
import {ActionImpactProps} from './types';

/**
 * Action Ã  impact du panier d'actions (carte + modale)
 */

export const ActionImpact = ({
  panier,
  onToggleSelected,
  ...props
}: ActionImpactProps) => {
  const [isActionSelected, setIsActionSelected] = useState(panier);

  useEffect(() => setIsActionSelected(panier), [panier]);

  const handleToggleSelect = value => {
    setIsActionSelected(value);
    onToggleSelected(value);
  };

  return (
    <ModaleActionImpact
      panier={isActionSelected}
      onToggleSelected={handleToggleSelect}
      {...props}
    >
      <div>
        <CarteActionImpact
          isSelected={isActionSelected}
          panier={isActionSelected}
          onToggleSelected={handleToggleSelect}
          {...props}
        />
      </div>
    </ModaleActionImpact>
  );
};
