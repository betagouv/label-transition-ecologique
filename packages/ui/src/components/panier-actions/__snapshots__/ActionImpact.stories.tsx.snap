// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Panier d'actions/Action à impact Default smoke-test 1`] = `
<div aria-expanded="false"
     aria-haspopup="dialog"
>
  <div class="p-7 m-px border bg-white rounded-lg flex flex-col gap-4 text-primary-9 text-base font-bold active:!bg-white group hover:cursor-pointer hover:shadow-card hover:border-grey-4 border-grey-3 box-content"
       style="max-width: 400px;"
  >
    <div class="text-primary-8 text-sm font-bold">
      <div class="flex justify-between">
        <div>
          Mieux consommer
        </div>
        <div>
          <span class="ri-money-euro-circle-fill font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] text-secondary-1">
          </span>
          <span class="ri-money-euro-circle-fill font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] text-secondary-1">
          </span>
          <span class="ri-money-euro-circle-fill font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] text-grey-4">
          </span>
        </div>
      </div>
    </div>
    Désigner une (ou plusieurs personnes) chef de projet transition écologique
    <div class="text-primary-8 text-sm font-bold mt-auto">
      <div class="flex justify-between items-center relative">
        <div data-test="Badge-undefined"
             class="bg-success-2 border-success-2 flex items-center gap-1 max-w-max h-fit px-3 py-1 border border-solid rounded !px-1.5 !py-0.5 absolute group-hover:hidden"
        >
          <span class="text-success-1 font-bold uppercase leading-4 text-left line-clamp-1 text-xs">
            Complexité simple
          </span>
        </div>
        <div class="flex gap-2 invisible group-hover:visible ml-auto">
          <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-primary-7 hover:!text-primary-8 bg-primary-0 hover:!bg-primary-1 border-primary-7 hover:!border-primary-8 text-xs py-2 px-5 !px-4">
            Non pertinent
          </button>
          <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-primary-7 hover:!text-primary-8 bg-primary-0 hover:!bg-primary-1 border-primary-7 hover:!border-primary-8 text-xs py-2 px-5 !px-4">
            En cours
          </button>
          <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-primary-7 hover:!text-primary-8 bg-primary-0 hover:!bg-primary-1 border-primary-7 hover:!border-primary-8 text-xs py-2 px-5 !px-4">
            Réalisé
          </button>
          <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-white hover:!text-white bg-primary-7 hover:!bg-primary-8 border-primary-7 hover:!border-primary-8 text-xs py-2 px-5 !px-4">
            Ajouter
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
